### PR TITLE
Handle exception if ConsoleHost tries to set cursor out of bounds because screen buffer changed

### LINKED ISSUE #15254: PowerShell process occasionally terminates due to Win32 internal error at SetConsoleCursorPosition
I'm seeing an exception terminating PowerShell when I run a script calling devenv.com inside. Because that happens occasionally, I want to understand if there is a similar known issue or how I can collect actionable logs.

## Steps to reproduce

Run a PowerShell script invoking DevEnv.com (v16.9.2) in Windows Terminal Preview 1.7.572.0. The script also calls `Write-Progress` as well as `Write-Information` before that.

## Expected behavior

The script always completes without terminating PowerShell process.

## Actual behavior

PowerShell process terminates. This happens once out of 10-20 times. An error message is output as follows. I must close the window and start a new PowerShell window.

```none
Process terminated. The Win32 internal error "The parameter is incorrect." 0x57 occurred while trying to set the cursor position. Contact Microsoft Customer Support Services.
   at System.Environment.FailFast(System.String, System.Exception)
   at Microsoft.PowerShell.UnmanagedPSEntry.Start(System.String[], Int32)
   at Microsoft.PowerShell.ManagedPSEntry.Main(System.String[])
System.Management.Automation.Host.HostException: The Win32 internal error "The parameter is incorrect." 0x57 occurred while trying to set the cursor position. Contact Microsoft Customer Support Services.
 ---> System.ComponentModel.Win32Exception (87): The parameter is incorrect.
   --- End of inner exception stack trace ---
   at Microsoft.PowerShell.ConsoleControl.SetConsoleCursorPosition(SafeFileHandle consoleHandle, Coordinates cursorPosition)
   at Microsoft.PowerShell.ConsoleHostRawUserInterface.set_CursorPosition(Coordinates value)
   at Microsoft.PowerShell.ProgressPane.Hide()
   at Microsoft.PowerShell.ConsoleHostUserInterface.ResetProgress()
   at Microsoft.PowerShell.Executor.ExecuteCommandHelper(Pipeline tempPipeline, Exception& exceptionThrown, ExecutionOptions options)
   at Microsoft.PowerShell.Executor.ExecuteCommand(String command, Exception& exceptionThrown, ExecutionOptions options)
   at Microsoft.PowerShell.ConsoleHost.InputLoop.Run(Boolean inputLoopIsNested)
   at Microsoft.PowerShell.ConsoleHost.InputLoop.RunNewInputLoop(ConsoleHost parent, Boolean isNested)
   at Microsoft.PowerShell.ConsoleHost.EnterNestedPrompt()
   at Microsoft.PowerShell.ConsoleHost.DoRunspaceLoop(String initialCommand, Boolean skipProfiles, Collection`1 initialCommandArgs, Boolean staMode, String configurationName)
   at Microsoft.PowerShell.ConsoleHost.Run(CommandLineParameterParser cpp, Boolean isPrestartWarned)
   at Microsoft.PowerShell.ConsoleHost.Start(String bannerText, String helpText)
   at Microsoft.PowerShell.UnmanagedPSEntry.Start(String[] args, Int32 argc)
[process exited with code 2148734499]
```

The exception is thrown on executing the following line.

```powershell
& DevEnv.com $sln /$Target "$Configuration|$Platform" /Project $Project | Tee-Object -Variable 'output'
```

## Environment data

```none
Name                           Value
----                           -----
PSVersion                      7.2.0-preview.4
PSEdition                      Core
GitCommitId                    7.2.0-preview.4
OS                             Microsoft Windows 10.0.19042
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### PR DESCRIPTION
**Pull request by SteveL-MSFT**
# PR Summary

There's a race condition between the call to set the console cursor position and the screen buffer size has changed that puts the coordinates outside of the screen buffer area and causes pwsh to crash with an unhandled exception.  Also, it seems that the console host is calling Win32 APIs instead of the .NET APIs and I don't see any specific reason to do that other than consistency with other native APIs that need to be used.

Fix is to switch to using Console.SetCursorPosition and remove all use of the Win32 API.  Catch the ArgumentOutOfRangeException and ignore it.  There is really no way to avoid the race condition here as any validation of the screen buffer size can be wrong by the time we set the cursor position.  The user may have resized the window or when tabs show up on Windows Terminal, the screen buffer size automatically shrinks so that the window size doesn't change.

## PR Context

Fix https://github.com/PowerShell/PowerShell/issues/15254

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` or `[ WIP ]` to the beginning of the title (the `WIP` bot will keep its status check at `Pending` while the prefix is present) and remove the prefix when the PR is ready.
- **[Breaking changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)**
    - [x] None
    - **OR**
    - [ ] [Experimental feature(s) needed](https://github.com/MicrosoftDocs/PowerShell-Docs/blob/staging/reference/6/Microsoft.PowerShell.Core/About/about_Experimental_Features.md)
        - [ ] Experimental feature name(s): 
- **User-facing changes**
    - [x] Not Applicable
    - **OR**
    - [ ] [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed: 
- **Testing - New and feature**
    - [x] N/A or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
- **Tooling**
    - [x] I have considered the user experience from a tooling perspective and don't believe tooling will be impacted.
    - **OR**
    - [ ] I have considered the user experience from a tooling perspective and opened an issue in the relevant tool repository. This may include:
        - [ ] Impact on [PowerShell Editor Services](https://github.com/PowerShell/PowerShellEditorServices) which is used in the [PowerShell extension](https://github.com/PowerShell/vscode-powershell) for VSCode
        (which runs in a different PS Host).
            - [ ] Issue filed: 
        - [ ] Impact on Completions (both in the console and in editors) - one of PowerShell's most powerful features.
            - [ ] Issue filed: 
        - [ ] Impact on [PSScriptAnalyzer](https://github.com/PowerShell/PSScriptAnalyzer) (which provides linting & formatting in the editor extensions).
            - [ ] Issue filed: 
        - [ ] Impact on [EditorSyntax](https://github.com/PowerShell/EditorSyntax) (which provides syntax highlighting with in VSCode, GitHub, and many other editors).
            - [ ] Issue filed:

### CODE REVIEW

File: src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs [hunk 1]
```diff
@@ -190,14 +190,15 @@ public override
...
-
-                ConsoleControl.CONSOLE_SCREEN_BUFFER_INFO bufferInfo;
-
-                ConsoleHandle handle = GetBufferInfo(out bufferInfo);
-
-                CheckCoordinateWithinBuffer(ref value, ref bufferInfo, "value");
-                ConsoleControl.SetConsoleCursorPosition(handle, value);
+                try
+                {
+                    Console.SetCursorPosition(value.X, value.Y);
```
Comment (iSazonov) [inline src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs:195]: If we now use Console class in Set why not use the class in Get too - Console.GetCursorPosition().

File: src/System.Management.Automation/utils/PInvokeDllNames.cs [hunk 2]
```diff
@@ -116,7 +116,6 @@ internal static class PinvokeDllNames
         internal const string PeekConsoleInputDllName = "api-ms-win-core-console-l2-1-0.dll";                /*103*/
         internal const string GetNumberOfConsoleInputEventsDllName = "api-ms-win-core-console-l1-1-0.dll";   /*104*/
         internal const string SetConsoleCtrlHandlerDllName = "api-ms-win-core-console-l1-1-0.dll";           /*105*/
-        internal const string SetConsoleCursorPositionDllName = "api-ms-win-core-console-l2-1-0.dll";        /*106*/
```
Comment (iSazonov) [inline src/System.Management.Automation/utils/PInvokeDllNames.cs:119]: Every line has a number in comment - make sense to update the numbers in the list tail?

**Comment by iSazonov** [comment]
Iwonder that tests fail.
.Net GetBufferInfo() implementation https://github.com/dotnet/runtime/blob/1b09a384f29eafd98ec6bcb2d9e6fc820c9db801/src/libraries/System.Console/src/System/ConsolePal.Windows.cs#L1066-L1077
PowerShell implementation looks the same"
https://github.com/PowerShell/PowerShell/blob/ae772462a5e9da61ab1858f46093abff83380c39/src/Microsoft.PowerShell.ConsoleHost/host/msh/ConsoleHostRawUserInterface.cs#L1290-L1296

**Comment by SteveL-MSFT** [comment]
@iSazonov it appears that the difference in the .NET code and the previous PS code is that PS is only using the stdout handle while .NET is checking stdin, stdout, and stderr and if any of them fail, an exception is thrown.  In this case, I suspect stdin is not a valid handle on CI.

Given that the .NET code is working differently here, I think we should revert the GetCursorPosition code back.

**Review by iSazonov — APPROVED** [review]

**Review by anmenaga — APPROVED** [review]
LGTM

**Pull request merged**